src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,888 @@
+
+        /// <summary>
+        /// Process id of backend server.
+        /// This can only be called when there is an active connection.
+        /// </summary>
+        public int ProcessID => _npgsqlConnection.Connector?.BackendProcessId ?? throw new InvalidOperationException($"The {ProcessID} property can only be used when there is an active connection");
+
+        #endregion Properties
+
+        #region OpenAsync / DisposeAsync
+
+        /// <summary>
+        /// Opens a database replication connection with the property settings specified by the
+        /// <see cref="NpgsqlReplicationConnection.ConnectionString">ConnectionString</see>.
+        /// </summary>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.
+        /// The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A task representing the asynchronous open operation.</returns>
+        public async Task OpenAsync(CancellationToken cancellationToken = default)
+        {
+            // No need for NoSynchronizationContextScope.Enter() since
+            // NpgsqlConnection.OpenAsync() does it itself and it's the
+            // first/only async method call.
+            using var openState = EnsureAndSetState(ReplicationConnectionState.Closed, ReplicationConnectionState.Opening);
+            await _npgsqlConnection.OpenAsync(cancellationToken);
+
+            SetTimeouts(CommandTimeout, CommandTimeout);
+
+            openState.ChangeResetState(ReplicationConnectionState.Idle);
+        }
+
+        /// <summary>
+        /// Closes the replication connection and performs tasks associated
+        /// with freeing, releasing, or resetting its unmanaged resources asynchronously.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous dispose operation.</returns>
+        public ValueTask DisposeAsync()
+        {
+            using (NoSynchronizationContextScope.Enter())
+                return DisposeAsyncInternal();
+        }
+
+        async ValueTask DisposeAsyncInternal()
+        {
+            var currentState =
+                (ReplicationConnectionState)Interlocked.Exchange(ref _state, (int)ReplicationConnectionState.Disposed);
+            if (currentState == ReplicationConnectionState.Disposed)
+                return;
+
+            // The expected sequence is probably to first cancel the streaming replication and then dispose
+            // the connection, but if DisposeAsync() is called while the connection is streaming, we
+            // do the cancellation ourselves to bring it down in a predictable way.
+            if (currentState == ReplicationConnectionState.Streaming)
+                Cancel();
+
+#if NETSTANDARD2_0
+            _sendFeedbackTimer.Dispose();
+            _requestFeedbackTimer.Dispose();
+#else
+            await _sendFeedbackTimer.DisposeAsync();
```

Comment:
There's a bit of weirdness here. It seems that DisposeAsync is the analog of [Dispose(WaitHandle)](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer.dispose?view=netcore-3.1#System_Threading_Timer_Dispose_System_Threading_WaitHandle_), and not of the regular [Dispose](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer.dispose?view=netcore-3.1#System_Threading_Timer_Dispose). In other words, DisposeAsync will wait for any outstanding callbacks to complete (like Dispose(WaitHandle)), rather than just deleting the timer and continuing without waiting.

I don't see a particular reason for us to wait for the callback to end, so I think we should just always call Dispose. If we do want to wait, then we must also do that above for NETSTANDARD2_0 (by calling Dispose(WaitHandle), which will block synchronously).

/cc @vonzshik, we just discussed this in another context.

Reply:
I see.

Reply:
https://github.com/Brar/npgsql/pull/10